hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestInsertIntoOperation.scala

```diff
@@ -0,0 +1,329 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.functional
+
+import org.apache.hudi.common.testutils.HoodieTestDataGenerator
+import org.apache.hudi.common.testutils.RawTripTestPayload.recordsToStrings
+import org.apache.spark.sql
+import org.apache.spark.sql.hudi.HoodieSparkSqlTestBase
+import org.apache.hudi.common.model.WriteOperationType
+import org.apache.hudi.common.table.HoodieTableMetaClient
+import org.apache.hudi.common.table.timeline.TimelineUtils
+
+import scala.collection.JavaConversions._
+
+@DisableCI
+class TestInsertIntoOperation extends HoodieSparkSqlTestBase {
+
+  /**
+   * asserts if number of commits = count
+   * returns true if last commit is bulk insert
+   */
+  def assertCommitCountAndIsLastBulkInsert(basePath: String, count: Int): Boolean = {
+    val metaClient = HoodieTableMetaClient.builder()
+      .setBasePath(basePath)
+      .setConf(spark.sessionState.newHadoopConf())
+      .build()
+    val timeline = metaClient.getActiveTimeline.getAllCommitsTimeline
+    assert(timeline.countInstants() == count)
+    val latestCommit = timeline.lastInstant()
+    assert(latestCommit.isPresent)
+    assert(latestCommit.get().isCompleted)
+    val metadata = TimelineUtils.getCommitMetadata(latestCommit.get(), timeline)
+    metadata.getOperationType.equals(WriteOperationType.BULK_INSERT)
+  }
+
+  def createTable(tableName: String, writeOptions: String, tableBasePath: String): Unit = {
+    spark.sql(
+      s"""
+         | create table $tableName (
+         |  timestamp long,
+         |  _row_key string,
+         |  rider string,
+         |  driver string,
+         |  begin_lat double,
+         |  begin_lon double,
+         |  end_lat double,
+         |  end_lon double,
+         |  fare STRUCT<
+         |    amount: double,
+         |    currency: string >,
+         |  _hoodie_is_deleted boolean,
+         |  partition_path string
+         |) using hudi
+         | partitioned by (partition_path)
+         | $writeOptions
+         | location '$tableBasePath'
+         |
+    """.stripMargin)
+  }
+
+  def generateInserts(dataGen: HoodieTestDataGenerator, instantTime: String, n: Int): sql.DataFrame = {
+    val recs = dataGen.generateInsertsNestedExample(instantTime, n)
+    spark.read.json(spark.sparkContext.parallelize(recordsToStrings(recs), 2))
+  }
+
+  def doInsert(dataGen: HoodieTestDataGenerator, tableName: String, instantTime: String): Unit = {
+    generateInserts(dataGen, instantTime, 100).select("timestamp", "_row_key", "rider", "driver",
+      "begin_lat", "begin_lon", "end_lat", "end_lon", "fare", "_hoodie_is_deleted", "partition_path").
+      createOrReplaceTempView("insert_temp_table")
+    spark.sql(s"insert into $tableName select * from insert_temp_table")
+  }
+
+  test("No configs set") {
```

Comment:
can we also test pk less table. 
both table types. 
no preCombine.

Reply:
yes, we might need to fix it. 
check the fix we have added here 
https://github.com/apache/hudi/pull/8879/files#diff-560283e494c8ba8da102fc217a2201220dd4db731ec23d80884e0f001a7cc0bc

for now, you can incorporate it in this patch and add tests.